srml/babe/src/lib.rs

```diff
@@ -207,65 +201,94 @@ impl<T: Trait> Module<T> {
-		<system::Module<T>>::digest()
+	/// Call this function exactly once when an epoch changes, to update the
+	/// randomness. Returns the new randomness.
+	fn randomness_change_epoch(epoch_index: u64) -> [u8; 32] {
+		let this_randomness = NextRandomness::get();
+		let next_randomness = compute_randomness(
+			this_randomness,
+			epoch_index,
+			UnderConstruction::get(),
+		);
+		UnderConstruction::put(&[0; 32]);
+		NextRandomness::put(&next_randomness);
+		this_randomness
 	}
+
 }
 
 impl<T: Trait> OnTimestampSet<T::Moment> for Module<T> {
 	fn on_timestamp_set(_moment: T::Moment) { }
 }
 
-impl<T: Trait> session::OneSessionHandler<T::AccountId> for Module<T> {
-	type Key = AuthorityId;
+pub trait Duration {
+	fn babe_epoch_duration() -> u64;
+}
 
-	fn on_new_session<'a, I: 'a>(changed: bool, validators: I)
+impl<T: Trait + staking::Trait + Duration> session::OneSessionHandler<T::AccountId> for Module<T> {
+	type Key = AuthorityId;
+	fn on_new_session<'a, I: 'a>(_changed: bool, _validators: I, queued_validators: I)
 		where I: Iterator<Item=(&'a T::AccountId, AuthorityId)>
 	{
-		// instant changes
-		if changed {
-			let next_authorities = validators.map(|(_, k)| k).collect::<Vec<_>>();
-			let last_authorities = <Module<T>>::authorities();
-			if next_authorities != last_authorities {
-				Self::change_authorities(next_authorities);
-			}
-		}
-
-		let rho = UnderConstruction::get();
-		UnderConstruction::put([0; 32]);
-		let last_epoch_randomness = EpochRandomness::get();
+		use staking::BalanceOf;
+		let to_votes = |b: BalanceOf<T>|
+			<T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b);
 		let epoch_index = EpochIndex::get()
 			.checked_add(1)
 			.expect("epoch indices will never reach 2^64 before the death of the universe; qed");
 		EpochIndex::put(epoch_index);
-		EpochRandomness::put(NextEpochRandomness::get());
-		let mut s = [0; 72];
-		s[..32].copy_from_slice(&last_epoch_randomness);
-		s[32..40].copy_from_slice(&epoch_index.to_le_bytes());
-		s[40..].copy_from_slice(&rho);
-		NextEpochRandomness::put(runtime_io::blake2_256(&s))
+
+		// *Next* epoch’s authorities.
+		let authorities = queued_validators.map(|(account, k)| {
+			(k, to_votes(staking::Module::<T>::stakers(account).total))
+		}).collect::<Vec<_>>();
+
+		// What was the next epoch is now the current epoch
+		let randomness = Self::randomness_change_epoch(epoch_index);
+		Self::change_epoch(Epoch {
+			randomness,
+			authorities,
+			epoch_index,
+			duration: T::babe_epoch_duration(),
+		})
 	}
 
 	fn on_disabled(i: usize) {
-		let log: DigestItem<T::Hash> = DigestItem::Consensus(
-			BABE_ENGINE_ID,
-			ConsensusLog::OnDisabled(i as u64).encode(),
-		);
-		<system::Module<T>>::deposit_log(log.into());
+		Self::deposit_consensus(ConsensusLog::OnDisabled(i as u64))
 	}
 }
 
+fn compute_randomness(
+	last_epoch_randomness: [u8; 32],
+	epoch_index: u64,
+	rho: [u8; VRF_OUTPUT_LENGTH],
+) -> [u8; 32] {
+	let mut s = [0; 40 + VRF_OUTPUT_LENGTH];
```

Comment:
this function is much more easily written `T::Hash::hash_of((last_epoch_randomness, epoch_index, rho))`